CI: Android Emulator Tests Failing Due to Disk Space Exhaustion

# CI: Android Emulator Tests Failing Due to Disk Space Exhaustion

## Problem

All Android emulator tests in CI are failing with a fatal disk space error. The emulator cannot start because there's insufficient disk space to create the userdata partition.

**Error:**
```
FATAL | Not enough space to create userdata partition. 
Available: 5215.37 MB at /home/runner/.android/avd/test.avd
Need: 7372.80 MB
```

**Result:** Emulator fails to boot → ADB cannot connect → Tests timeout after 10 minutes

## Impact

- ❌ JUnit Runner Emulator Tests - FAILING
- ❌ Playground Automobile Emulator Tests - FAILING
- All PRs with emulator tests are blocked
- Retrying jobs doesn't help (disk space issue persists)

## Timeline

**Last successful run:** 2026-01-09 15:05:27Z (PR #459)
**First failure:** 2026-01-09 15:32:40Z (PR #520, after #517 merged)

## Root Cause

**PR #499** ("Cache Android system images to reduce CI setup time") added system image caching with an incorrect path:

```yaml
# .github/actions/android-emulator/action.yml:67
~/.android/sdk/system-images/*  # ❌ Wrong path
```

**Actual SDK location:** `/usr/local/lib/android/sdk/system-images/`

### Why This Causes Disk Space Issues

1. The incorrect path means system images aren't actually being cached
2. Even with the wrong path, the cache action may consume disk space trying
3. Each emulator run needs ~7.4 GB for userdata partition
4. Runner only has ~5.2 GB available after other operations
5. System images are re-downloaded each time (not cached), consuming additional space

## Evidence

### Logs from Failed Run (20856885286)

**Emulator startup:**
```
INFO | Found systemPath /usr/local/lib/android/sdk/system-images/android-35/google_apis/x86_64/
FATAL | Not enough space to create userdata partition. Available: 5215.37 MB, need 7372.80 MB
```

**ADB connection attempts (300+ retries):**
```
[command]/usr/local/lib/android/sdk/platform-tools/adb -s emulator-5554 shell getprop sys.boot_completed
adb: device 'emulator-5554' not found
The process '/usr/local/lib/android/sdk/platform-tools/adb' failed with exit code 1
```

**Final timeout:**
```
##[error]Timeout waiting for emulator to boot.
```

### Affected Jobs
- https://github.com/kaeawc/auto-mobile/actions/runs/20856885286/job/59926351781
- https://github.com/kaeawc/auto-mobile/actions/runs/20856885286/job/59926351792

## Proposed Solution

Implement both Fix 2 and Fix 3 together:

### Fix 2: Correct the System Image Cache Path

Update `.github/actions/android-emulator/action.yml`:

```yaml
- name: AVD cache
  uses: actions/cache@v4
  id: avd-cache
  with:
    path: |
      ~/.android/avd/*
      ~/.android/adb*
      /usr/local/lib/android/sdk/system-images/*  # ✅ Corrected path
    key: avd-${{ runner.os }}-${{ inputs.api_level }}-${{ inputs.arch }}-${{ inputs.target }}
    restore-keys: |
      avd-${{ runner.os }}-${{ inputs.api_level }}-${{ inputs.arch }}-
      avd-${{ runner.os }}-${{ inputs.api_level }}-
      avd-${{ runner.os }}-
```

**Benefits:**
- Actually caches system images (speeds up CI)
- Reduces download bandwidth
- Maintains the optimization goal from PR #499

### Fix 3: Add Disk Cleanup Before Emulator Tests

Add cleanup step to free space before emulator operations:

```yaml
- name: Free disk space for emulator
  shell: ${{ inputs.shell }}
  run: |
    echo "Disk space before cleanup:"
    df -h
    
    # Remove large pre-installed tools that aren't needed
    sudo rm -rf /usr/share/dotnet
    sudo rm -rf /opt/ghc
    sudo rm -rf /usr/local/share/boost
    sudo rm -rf "$AGENT_TOOLSDIRECTORY"
    
    echo ""
    echo "Disk space after cleanup:"
    df -h
```

**Insert location:** After "Enable KVM group perms", before "AVD cache"

**Benefits:**
- Frees ~10-15 GB of disk space
- Standard practice for Android emulator CI
- Ensures enough space for userdata partition (7.4 GB needed)

## Implementation Plan

1. **Create PR with both fixes**
   - Fix cache path in `.github/actions/android-emulator/action.yml`
   - Add disk cleanup step
   - Test on a failing PR to verify fix

2. **Verify cache is working**
   - First run: Cache miss (downloads system images) - slower
   - Second run: Cache hit (uses cached images) - fast
   - Check cache size in Actions artifacts

3. **Monitor disk usage**
   - Add `df -h` output before/after key steps
   - Ensure cleanup provides sufficient space
   - Verify emulator boots successfully

## Testing Approach

1. Apply fixes to a branch
2. Trigger emulator tests (push or manual workflow dispatch)
3. Verify:
   - ✅ Disk cleanup frees sufficient space
   - ✅ System images cached at correct path
   - ✅ Emulator boots successfully
   - ✅ ADB connects
   - ✅ Tests run and pass
4. Check second run for cache hit (faster CI time)

## Alternative: Quick Fix (Remove Cache)

If we need to unblock CI immediately, we could remove the system image cache line entirely:

```diff
  path: |
    ~/.android/avd/*
    ~/.android/adb*
-   ~/.android/sdk/system-images/*
```

**Pros:** Immediate fix, simple
**Cons:** Loses caching optimization, slower CI times

**Recommendation:** Go with the comprehensive solution (Fix 2 + Fix 3) to maintain performance while fixing the issue.

## Related PRs

- #499 - Initial cache optimization (introduced incorrect path)
- #453 - Previous emulator CI fixes
- #451, #467 - Prior disk cleanup for other jobs

## Acceptance Criteria

- [ ] Emulator tests pass consistently in CI
- [ ] System images are cached at correct path
- [ ] Disk cleanup provides sufficient space (verify with `df -h`)
- [ ] Cache hit on subsequent runs (verify in logs)
- [ ] No timeout errors
- [ ] Both JUnit and Playground emulator tests succeed

--- a/.github/actions/android-emulator/action.yml
+++ b/.github/actions/android-emulator/action.yml
@@ -67,7 +67,6 @@ runs:
         sudo rm -rf /usr/share/dotnet
         sudo rm -rf /opt/ghc
         sudo rm -rf /usr/local/share/boost
-        sudo rm -rf "$AGENT_TOOLSDIRECTORY"
 
         echo ""
         echo "Disk space after cleanup:"

